GoogleForCreators/web-stories-wp
# Build packages using Rollup

## Context



Before we can publish our packages on npm we need a build step for them, ideally to produce both CommonJS and ESM versions.

## Summary



This is an attempt at this by leveraging Rollup and several Rollup plugins that fit our needs.

Building packages creates `dist` and `dist-module` folders that are then used for distribution.

The `src` folder is entirely ignored for publishing on npm.

Basic usage:

```bash
# Build all packages using Rollup
$ npm run bundle

# Build only the migration package
$ npm run bundle -- --configPackages=migration

# Build only the module version of all packages
$ npm run bundle -- --configEntries=es

# Build only the CommonJS version of the migration and text-sets packages
$ npm run bundle -- --configEntries=cjs --configPackages=migration,text-sets
```

## Relevant Technical Choices



Decided to try Rollup because we're already using it in the project and it was relatively easy to get off the ground.

Changed some imports to avoid circular dependencies, though this still needs more work.

Changed some imports to avoid mixing default imports and named exports. Quote:

> It is bad practice to mix default and named exports in the same module, though it is allowed by the specification.

Used custom resolvers for webpack and Jest so they continue to look for packages in the `src` folders instead of `dist`.

## To-do



* [x] Fix inline web worker usage in `story-editor`
* [x] Fix `import/no-unresolved` false positives
* [ ] Add sourcemaps?
* [ ] Fix duplicate React version in `react-photo-gallery`
* [ ] Fix some circular dependencies (see below)

```
(!) Circular dependencies
packages/story-editor/src/masks/index.js -> packages/story-editor/src/elements/index.js -> packages/story-editor/src/elements/elementTypes.js -> packages/story-editor/src/elements/text/index.js -> packages/story-editor/src/elements/text/display.js -> packages/story-editor/src/elements/shared/index.js -> packages/story-editor/src/utils/elementBorder.js -> packages/story-editor/src/masks/index.js

packages/story-editor/src/app/story/index.js -> packages/story-editor/src/app/story/useStoryReducer/reducers/index.js -> packages/story-editor/src/app/story/useStoryReducer/reducers/combineElements.js -> packages/story-editor/src/masks/index.js -> packages/story-editor/src/elements/index.js -> packages/story-editor/src/elements/elementTypes.js -> packages/story-editor/src/elements/text/index.js -> packages/story-editor/src/elements/text/edit.js -> packages/story-editor/src/app/story/index.js

packages/story-editor/src/app/story/index.js -> packages/story-editor/src/app/story/useStoryReducer/reducers/index.js -> packages/story-editor/src/app/story/useStoryReducer/reducers/combineElements.js -> packages/story-editor/src/masks/index.js -> packages/story-editor/src/elements/index.js -> packages/story-editor/src/elements/elementTypes.js -> packages/story-editor/src/elements/text/index.js -> packages/story-editor/src/elements/text/frame.js -> packages/story-editor/src/app/story/index.js

...and 48 more

```

## User-facing changes



None

## Testing Instructions






This PR can be tested by following these steps:

1. Verify that all status checks pass
2. Run `npm run bundle` to build the packages yourself


## Reviews

### Does this PR have a security-related impact?



Not that I could think of

### Does this PR change what data or activity we track or use?



No

### Does this PR have a legal-related impact?



No

## Checklist




---



See #10251
